Addition of LevelZero metric support to the LevelZero device pool and IO Group

lhlawson opened the pull request:
- Relates to #2148 feature quest from github issues
- Provides Intel hardware support analogous to the NVIDIA support added in #1917 and #1914

The addition of ZET metric polling capability to the LevelZero IO Group, Device Pool, and API shim.   

General Metric Streamer flowchart implemented follows the LevelZero approach shown [here](https://spec.oneapi.io/level-zero/latest/_images/tools_metric_streamer.png)

Details of change:
- Updated LevelZero.cpp to set ZET environment variables and follow the general time based metric query approach outlined in the [LevelZero Tools API](https://spec.oneapi.io/level-zero/latest/tools/PROG.html#metrics). 
- Added querying for the 'ComputeBasics' group of Intel Hardware.  An additional metric group may be queried but is not supported at this time.  

Additional Tasks:
- [x] Define Default LevelZero Tools report/event rate based on workload performance
- [x] Add Test for signal and control of LevelZero Tools.
- [ ] Later PR: Consider adding LEVELZERO::METRICS:XVE_*:AGGREGATION.  Currently all signals are averaged over the number of reports received, but this could be modified to be AVG, LAST, MIN, MAX, etc
- [x] Test dynamic tracing with geopm controller (read_signal and batch read have been tested, but not with the controller) 
- [x] Add multi-tile support
- [x] Re-implement metric_destroy/teardown mechanism
- [x] Remove geopmread functionality until an appropriate tear-down handling with read_batch and read_signal can occur.
- [ ] Later PR: Add 'check and fail' if the metric group is changed out  from under the user

lhlawson committed:
First Pass LevelZero Tools & Metrics Support using metriccalcexp calls

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
Moved to single tile API, with GPU_CHIP support

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
Removed timing code

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
Cleanup, fixed debug message

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        if(m_devices.at(l0_device_idx).subdevice.m_metric_data.at(l0_domain_idx).count(metric_name) == 0) {
```
nit:
```suggestion
        if (m_devices.at(l0_device_idx).subdevice.m_metric_data.at(l0_domain_idx).count(metric_name) == 0) {
```

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+                            m_devices.at(l0_device_idx).subdevice.metric_streamer.at(l0_domain_idx));
+            }
+            else {
+                metric_read(l0_device_idx, l0_domain_idx); //TODO: possible infinite loop
```
:o 
Please address this when moving out of draft

avilcheslopez commented on service/src/LevelZero.hpp:
```diff
@@ -249,8 +249,23 @@ namespace geopm
+            virtual void metric_init(unsigned int l0_device_idx,
+                                     unsigned int l0_domain_idx) = 0;
+            virtual void metric_destroy(unsigned int l0_device_idx,
+                                        unsigned int l0_domain_idx) = 0;
```
Are these methods really needed? I see the init is being handled automatically internally, which means we can do without temporal couplings. I do see the code is still in draft form as the `metric_destroy()` implementation is commented out. However, if it is needed, I'm thinking it could be managed similarly how the init is being done.

lhlawson replied:
These have been removed int he most recent commit, however metric_destroy may be re-added at some point if we determine we want to be able to stop using ZET dynamically as discussed in the arch meeting

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+    void LevelZeroImp::metric_group_cache(unsigned int device_idx) {
+        for (int subdevice_idx = 0;
+         subdevice_idx < m_devices.at(device_idx).m_num_subdevice;
+         ++subdevice_idx) {
```
nit: please fix indentation

avilcheslopez commented on service/src/LevelZeroImp.hpp:
```diff
@@ -114,9 +128,25 @@ namespace geopm
+                std::vector<zet_metric_group_handle_t> metric_group_handle; //compute basic only
+
+                // required for L0 metric result tracking
+                mutable std::vector<std::map<std::string, std::vector<double>>> m_metric_data;
```
Following the convention I'm seeing in the rest of our codebase, we're not prefixing properties in these inner structs with "m_".

avilcheslopez commented on service/src/LevelZeroImp.hpp:
```diff
@@ -114,9 +128,25 @@ namespace geopm
+                ze_driver_handle_t driver;
                 zes_device_handle_t device_handle;
                 ze_device_properties_t property;
                 uint32_t m_num_subdevice;
```
Same comment as above regarding "m_" prefix. I know it's not part of this review, but it would be something very simple to make the code consistent.

avilcheslopez commented on service/src/LevelZeroImp.hpp:
```diff
@@ -114,9 +128,25 @@ namespace geopm
+                std::vector<uint32_t> num_reports;
+                std::vector<bool> metric_domain_cached;
+                std::vector<ze_event_pool_handle_t> event_pool;
+                std::vector<ze_event_handle_t> event; //TODO: rename metric_notification_event?
```
Sure

dannosliwcd commented on service/src/LevelZero.cpp:
```diff
@@ -19,11 +19,13 @@
 static void __attribute__((constructor)) geopm_levelzero_init(void)
 {
     setenv("ZES_ENABLE_SYSMAN", "1", 1);
+//    setenv("ZET_ENABLE_METRICS", "1", 1);
```
```suggestion
```

lhlawson replied:
Re-enabled ZET by default.

dannosliwcd commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+        for (int subdevice_idx = 0;
+         subdevice_idx < m_devices.at(device_idx).m_num_subdevice;
+         ++subdevice_idx) {
+            //assume false
```
Seems to just be truth rather than assumption at this point, since it isn't cached or initialized yet

dannosliwcd commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+
+                std::vector<zet_metric_group_handle_t> metric_group_handle(num_metric_group);
+
+                //TODO: save the relevant per GPU pointers
```
Which pointers?

lhlawson replied:
TODO removed, this was a hold over from when the metric group handle, context, etc were not being stored in the tracking struct.

dannosliwcd commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+                                __LINE__);
+
+                //sampling period in nanoseconds
+                m_devices.at(device_idx).metric_sampling_period = 2000000;
```
Optionally rename to make sure the units are known everywhere this is used. E.g., `sampling_period_ns`

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+            m_devices.at(device_idx).subdevice.metrics_initialized.push_back(false);
+
+            char *zet_enable_metrics = getenv("ZET_ENABLE_METRICS");
+            if (zet_enable_metrics == NULL || strcmp(zet_enable_metrics, "1") != 0) {
```
Avoid use of `strcmp()`. Alternatives:
- Convert `char*` to `std::string` and use `std::string`'s `==` operator.
- Use `strncmp()`.

I know you're only comparing one character (and you first check `zet_enable_metrics` is not null), so there is no risk of `strcmp()` going out of bounds, so this comment is more for formality and to prevent future changes to inadvertently introduce an issue.

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+                std::cerr << "Warning: <geopm>: ZET_ENABLE_METRICS not set to 1.  Skipping metric caching" <<
+                             " for device " << std::to_string(device_idx) << " subdevice " <<
+                             std::to_string(subdevice_idx) << std::endl;
+#endif
```
If `GEOPM_DEBUG` is not defined, this will result in an empty block for the true branch of this `if`. Rework this to avoid that situation. You could change the condition such that you invert the branches and have this warning be printed in the `else` branch.

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+            m_devices.at(device_idx).subdevice.metrics_initialized.push_back(false);
+
+            char *zet_enable_metrics = getenv("ZET_ENABLE_METRICS");
+            if (zet_enable_metrics == NULL || strcmp(zet_enable_metrics, "1") != 0) {
```
```suggestion
            if (zet_enable_metrics == nullptr || strcmp(zet_enable_metrics, "1") != 0) {
```

lhlawson replied:
Removed in the most recent commit, ZET_ENABLE_METRICS is set by default as part of the GEOPM run

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+
+                     //Confirm metric groups of interest exist
+                     if (metric_group_properties.samplingType == ZET_METRIC_GROUP_SAMPLING_TYPE_FLAG_TIME_BASED
+                         && strcmp(metric_group_properties.name, "ComputeBasic") == 0) {
```
Same comment as before regarding `strcmp()`.

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
             }
         }
     }
+    void LevelZeroImp::metric_group_cache(unsigned int device_idx) {
```
I would recommend you rename this method to something that describes what the method does. After reading the code, it seems like this method is initializing the metric group cache. Is that correct?

I just noticed you're following the same naming pattern as the other methods that perform similar cache initialization. At this point, it might be too much trouble to rename all the methods for consistency. So please keep this in mind for the impending refactor of this code and in general for method naming.

lhlawson replied:
How does this look:

- metric_group_cache -> metric_group_init 
- metric_init -> metric_execute

My main concern would be confusion of our metric group initialization vs the ZET concept of initializing the metric group.

dannosliwcd commented on service/src/LevelZeroImp.hpp:
```diff
@@ -155,6 +188,10 @@ namespace geopm
             std::vector<ze_driver_handle_t> m_levelzero_driver;
             std::vector<m_device_info_s> m_devices;
+
+            void metric_group_cache(unsigned int l0_device_idx);
```
Need a comment explaining what this is caching and/or what data or functions are invalid until this is cached. My read-through understanding is:
* Caches to several parts of m_devices. I guess this function needs to be called after initialization and before attempting to request any metric data.
* Caches the names of metrics in the ComputeBasic L0 metric group. 

Maybe rename this to indicate that it is hard-coded to computebasic. Or parameterize the computebasic part.

dannosliwcd replied:
Or docs for the other functions might cover that (e.g., they say they'll throw if not yet cached)

lhlawson replied:
Comment added

dannosliwcd commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+        }
+    }
+
+    void LevelZeroImp::metric_destroy(unsigned int l0_device_idx, unsigned int l0_domain_idx)
```
Unused (unless that's just temporary)

dannosliwcd commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+        ze_result_t ze_result;
+        ze_context_handle_t context = m_devices.at(l0_device_idx).subdevice.context.at(l0_domain_idx);
+        ze_result = zetContextActivateMetricGroups(context, m_devices.at(l0_device_idx).subdevice_handle.at(l0_domain_idx),
+                                                   1, &m_devices.at(l0_device_idx).subdevice.metric_group_handle.at(l0_domain_idx));
```
Maybe comment 1 since this is initialized for computebasic only

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
             }
         }
     }
+    void LevelZeroImp::metric_group_cache(unsigned int device_idx) {
```
On a separate note about this method, I feel it's a bit lengthy and requires drilling down into nested blocks, so it's not straightforward to quickly grok what it's doing, so if you can, please break it down a bit.

dannosliwcd commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+        ze_event_pool_handle_t event_pool_handle = nullptr;
+        ze_event_pool_desc_t event_pool_desc = {ZE_STRUCTURE_TYPE_EVENT_POOL_DESC, nullptr, 0, 1};
+
+        ze_result = zeEventPoolCreate(context, &event_pool_desc, 1,
```
Knowledge check for myself. Docs say 1 here is the number of devices. Is this 1 because we're initializing one GPU at a time?

dannosliwcd replied:
From reading further down I guess it also means we're sampling one device at a time, since this is the event that read attempts block on while waiting for metric data

lhlawson replied:
We're targeting 1 event in each event pool, and 1 sub-device per event.

dannosliwcd commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+        zet_metric_streamer_desc_t metric_streamer_desc = {
+            ZET_STRUCTURE_TYPE_METRIC_STREAMER_DESC,
+            nullptr,
+            32768,
```
Why this number of reports per notification?

lhlawson replied:
Updated to 4 reports (1ms per report target)

dannosliwcd commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+        std::vector<uint32_t> metric_count(data_count);
+        std::vector<zet_typed_value_t> metric_values(num_metric_values);
+        ze_result = zetMetricGroupCalculateMultipleMetricValuesExp(m_devices.at(l0_device_idx).subdevice.metric_group_handle.at(l0_domain_idx), calculation_type, data_size, data.data(), &data_count, &num_metric_values, metric_count.data(), metric_values.data());
+//        check_ze_result(ze_result, GEOPM_ERROR_RUNTIME,
+//                        "LevelZero::" + std::string(__func__) +
+//                        ": LevelZero Metric group calculate metric values to calculate data failed",
+//                        __LINE__);
+//        //std::cout << "\tmetric group calculate metric values: " << std::to_string(num_metric_values) << std::endl;
+//        ze_result = ZE_RESULT_ERROR_UNKNOWN;
```
```suggestion
```

Looks like you removed this since you added a different check a few lines later

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+        zet_metric_streamer_desc_t metric_streamer_desc = {
+            ZET_STRUCTURE_TYPE_METRIC_STREAMER_DESC,
+            nullptr,
+            32768,
```
How did you come up with this number? I do see it's the max + 1 of a 2-byte int type, but why choose that number? Is that the max value for that descriptor's property? If you do actually need the max int value, use `<limits>`.

lhlawson replied:
Updated to 4.

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+
+    void LevelZeroImp::metric_calc(unsigned int l0_device_idx, unsigned int l0_domain_idx,
+                                   zet_metric_streamer_handle_t metric_streamer) const
+    {
```
Does this method require `metric_init()` and `metric_group_cache()` to already be called? If so, I would recommend adding an assertion (`GEOPM_DEBUG_ASSERT`) to ensure this.

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+                std::string metric_name (metric_properties.name);
+
+                //TODO: check timing impact of only processing metrics supported by IOGroup
+                if(metric_name.compare("XVE_ACTIVE") == 0 ||
```
nit:
```suggestion
                if (metric_name.compare("XVE_ACTIVE") == 0 ||
```

avilcheslopez replied:
Looking at the documentation, it seems `std::string::compare()` (without specifying the number of characters to compare) has the same vulnerability/risk as `strcmp()`. Please use `operator==()`.

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+
+                //TODO: check timing impact of only processing metrics supported by IOGroup
+                if(metric_name.compare("XVE_ACTIVE") == 0 ||
+                   metric_name.compare("XVE_STALL") == 0) {
+                    for (unsigned int report_idx = 0; report_idx < num_reports; report_idx++)
```
Fix indentation

dannosliwcd commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+                            m_devices.at(l0_device_idx).subdevice.metric_streamer.at(l0_domain_idx));
+            }
+            else {
+                metric_read(l0_device_idx, l0_domain_idx); //TODO: possible infinite loop
```
Should this have a timeout mechanism? E.g., change it to a loop around line 858 and trigger an error if too many attempts or too much time has passed without the event

dannosliwcd replied:
Might want to add a non-zero timeout to the last arg in zeEventHostSynchronize if going that route

lhlawson replied:
The recursive call is removed in the most recent version.  In cases where the event is not ready the old data will be provided to the user.
We may want to add a 'collected' timestamp so the user can identify if this happens.

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+                        //This is the actual gathering of the data
+                        zet_typed_value_t data = metric_values.at(report_idx*num_metric + metric_idx);
+                        double data_double = NAN;
+                        switch( data.type )
```
nit:
```suggestion
                        switch (data.type)
```

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+                            break;
+                        };
+                        //Clear cached values
+                        if(report_idx == 0) {
```
nit:
```suggestion
                        if (report_idx == 0) {
```

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+                        zet_typed_value_t data = metric_values.at(report_idx*num_metric + metric_idx);
+                        double data_double = NAN;
+                        switch( data.type )
+                        {
+                        case ZET_VALUE_TYPE_UINT32:
+                            data_double = data.value.ui32;
+                            break;
+                        case ZET_VALUE_TYPE_UINT64:
+                            data_double = data.value.ui64;
+                            break;
+                        case ZET_VALUE_TYPE_FLOAT32:
+                            data_double = data.value.fp32;
+                            break;
+                        case ZET_VALUE_TYPE_FLOAT64:
+                            data_double = data.value.fp64;
+                            break;
+                        case ZET_VALUE_TYPE_BOOL8:
+                            data_double = data.value.ui32;
+                            break;
+                        default:
+                            break;
+                        };
```
Does the API provide a helper function to do this? If not, consider adding a helper function to do this. If it's not needed anywhere else it would at least reduce this function's size and make it more readable.

lhlawson replied:
The API does not, I've moved it to metric_data_convert(...)

dannosliwcd commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+
+            ze_result_t ze_host_result = zeEventHostSynchronize(m_devices.at(l0_device_idx).subdevice.event.at(l0_domain_idx), 0);
+
+            if (ze_host_result != ZE_RESULT_NOT_READY) {
+                metric_calc(l0_device_idx, l0_domain_idx,
+                            m_devices.at(l0_device_idx).subdevice.metric_streamer.at(l0_domain_idx));
+            }
+            else {
```
Should probably be something like `if == NOT_READY then retry else if SUCCESS then calc() else error`

lhlawson replied:
Retry logic removed in most recent version.

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+                    for (unsigned int report_idx = 0; report_idx < num_reports; report_idx++)
+                    {
+                        //This is the actual gathering of the data
+                        zet_typed_value_t data = metric_values.at(report_idx*num_metric + metric_idx);
```
Preferably, it would be better if a data structure was used that more naturally represented the data being stored. If it's not convenient for whatever reason, please add a comment explaining this formula and how `metric_values` is structured such that this formula correctly indexes into the data item of interest.

I'm a bit nervous that `report_idx` depends on `num_reports` which is `num_metric_values / num_metric`. Does this formula take into account what happens when `num_metric_values` is not divisible by `num_metric`? If needed, please add checks and/or assertions to ensure any preconditions needed for this formula to work.

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -523,6 +526,366 @@ namespace geopm
+            }
+        }
+        else {
+            m_devices.at(l0_device_idx).subdevice.m_metric_data.at(l0_domain_idx).at("XVE_ACTIVE") = {};
+            m_devices.at(l0_device_idx).subdevice.m_metric_data.at(l0_domain_idx).at("XVE_STALL") = {};
```
Consider putting these strings (the metric names) in symbolic constants, as they're used in multiple places (including the IOGroup).

avilcheslopez commented on service/src/LevelZeroImp.hpp:
```diff
@@ -114,9 +128,25 @@ namespace geopm
+                std::vector<zet_metric_group_handle_t> metric_group_handle; //compute basic only
+
+                // required for L0 metric result tracking
+                mutable std::vector<std::map<std::string, std::vector<double>>> m_metric_data;
```
Q: why does this need the mutable specifier?

lhlawson replied:
It doesn't anymore!

lhlawson committed:
Fix-up from debug with Brad, disable all but 1 GPU

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
Addressed Review Feedback

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
Addressed Review Comments: function names, spacing, DEBUG asserts

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
fixup: missed change of metric_group_cache to metric_group_init

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
updated L0 mocks with ZET related functions

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
Addressed Review Feedback: remove m_ from inner members of structs

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
Legacy tests passing

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
Update read batch test fix num_report scaling factor issue

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
Added restriction on ZET read_signal, updated read_batch test

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
Added ZET tests for L0 dev pool, updated dev pool to use reduce

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

bgeltz commented on service/src/LevelZero.cpp:
```diff
@@ -13,17 +13,20 @@
 static void __attribute__((constructor)) geopm_levelzero_init(void)
 {
     setenv("ZES_ENABLE_SYSMAN", "1", 1);
+    setenv("ZET_ENABLE_METRICS", "1", 1);
```
Should this be set conditionally somehow?

If it's too late to set this when push_signal() is called with metrics based signals/controls, then I think this has to be done at compile time with a #define.

lhlawson replied:
We're setting this at all times, but having the actual usage of ZET (and context running on the GPU) be conditional on the user doing a push signal of a ZET metric and the first read of those metrics.

bgeltz commented on service/src/LevelZero.hpp:
```diff
@@ -249,8 +249,17 @@ namespace geopm
+
+            //Level Zero Tools/ZET
+            virtual std::vector<double> metric_sample(unsigned int l0_device_idx,
+                                                      unsigned int l0_domain_idx,
+                                                      std::string metric_name) const = 0;
+            virtual uint32_t metric_update_rate(unsigned int l0_device_idx) const = 0;
+
+            virtual void metric_read(unsigned int l0_device_idx,
+                                     unsigned int l0_domain_idx) = 0;
     };
 
-    const LevelZero &levelzero();
+    LevelZero &levelzero();
```
Why was this required?

lhlawson replied:
This change was required for the lazy init approach being used to update some level zero metric containers.  After more recent code changes it might not be required.

bgeltz commented on service/src/LevelZero.cpp:
```diff
@@ -178,7 +181,18 @@ namespace geopm
+    }
+
+    LevelZeroImp::~LevelZeroImp() {
+        for (unsigned int gpu_idx = 0; gpu_idx < m_num_gpu; gpu_idx++) {
```
```suggestion
        for (unsigned int gpu_idx = 0; gpu_idx < m_num_gpu; ++gpu_idx) {
```

Please be consistent with the variable incrementing.

bgeltz commented on service/src/LevelZero.cpp:
```diff
@@ -524,6 +538,425 @@ namespace geopm
+                            __LINE__);
+
+            // set metric_notifcation_event sampling period in nanoseconds
+            m_devices.at(device_idx).metric_sampling_period_ns =   1000000;
```
```suggestion
            m_devices.at(device_idx).metric_sampling_period_ns = 1000000; // 1 ms
```

bgeltz commented on service/src/LevelZero.cpp:
```diff
@@ -524,6 +538,425 @@ namespace geopm
+                   m_devices.at(device_idx).subdevice.num_metric.push_back(num_metric);
+
+                   //Build metric map
+                   for (unsigned int metric_idx = 0; metric_idx < num_metric; metric_idx++)
```
```suggestion
                   for (unsigned int metric_idx = 0; metric_idx < num_metric; ++metric_idx)
```

bgeltz commented on service/src/LevelZero.cpp:
```diff
@@ -524,6 +538,425 @@ namespace geopm
+                       m_devices.at(device_idx).subdevice.metric_data.at(subdevice_idx)[metric_name] = {};
+                       m_devices.at(device_idx).subdevice.metric_data.at(subdevice_idx)["NUM_REPORTS"] = {};
+                   }
+                   // Break out of loop once we've found the group of interest
+                   break;
```
```suggestion
```
This ```break``` is inside the if() block that starts on line 612.  I don't believe this was intentional.

lhlawson replied:
The intent is to break out of the for loop surrounding the if once we fine the ComputeBasics group that allows for ZET_METRIC_GROUP_SAMPLING_TYPE_FLAG_TIME_BASED sampling

lhlawson committed:
Apply suggestions from code review

Co-authored-by: Brad Geltz <brgeltz@gmail.com>

lhlawson commented on service/src/LevelZero.cpp:
```diff
@@ -524,6 +538,425 @@ namespace geopm
+                       m_devices.at(device_idx).subdevice.metric_data.at(subdevice_idx)[metric_name] = {};
+                       m_devices.at(device_idx).subdevice.metric_data.at(subdevice_idx)["NUM_REPORTS"] = {};
+                   }
+                   // Break out of loop once we've found the group of interest
```
```suggestion
                   // Break out of the metric group for loop once we've found the group of interest (ComputeBasic, time based sampling).  
```

lhlawson committed:
Add comment on for loop break

lhlawson committed:
Fixed typo

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
Made initial data sizing for metric stream a one time operation

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
removed dumping of extra data in buffer, removed error on L0 IOGroup read of METRIC signal

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

dannosliwcd commented on service/src/LevelZero.cpp:
```diff
@@ -702,167 +737,203 @@ namespace geopm
             nullptr,
-            32768,
-            m_devices.at(l0_device_idx).metric_sampling_period};
+            4, // number of reports to notify on.
```
Same question as before: why 4 reports at 1 ms per sample? Is that something to do with our typical 5ms agent loops?

lhlawson commented on service/src/LevelZero.cpp:
```diff
@@ -524,6 +538,426 @@ namespace geopm
+        zet_metric_streamer_desc_t metric_streamer_desc = {
+            ZET_STRUCTURE_TYPE_METRIC_STREAMER_DESC,
+            nullptr,
+            4, // number of reports to notify on.
```
```suggestion
            4, // number of reports to notify on.  Targeting 4 reports, 1 per millisecond as that will, generally, 
                // fall within the 5ms control loop and is a good tradeoff in terms of overhead vs visibility within the sampling period.
```

lhlawson committed:
Update comment on number of report choice.

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -524,6 +538,427 @@ namespace geopm
+        // Allocate memory for future reads
+        size_t data_size = 0;
+        ze_result = zetMetricStreamerReadData(m_devices.at(l0_device_idx).subdevice.metric_streamer.at(l0_domain_idx),
+                                              UINT32_MAX, &data_size, nullptr); //TODO: this value should match the report_count_req in metric_read
```
nit: alignment

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -178,7 +181,18 @@ namespace geopm
+        for (unsigned int gpu_idx = 0; gpu_idx < m_num_gpu; ++gpu_idx) {
+            for (int subdevice_idx = 0;
+             subdevice_idx < m_devices.at(gpu_idx).num_subdevice;
+             ++subdevice_idx) {
```
nit: fix indentation

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -524,6 +538,427 @@ namespace geopm
+
+            std::string metric_name (metric_properties.name);
+
+            //TODO: check timing impact of only processing metrics supported by IOGroup
```
TODO for this PR?

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -524,6 +538,427 @@ namespace geopm
+                            ": LevelZero Metric property acquisition failed",
+                            __LINE__);
+
+            std::string metric_name (metric_properties.name);
```
nit:
```suggestion
            std::string metric_name(metric_properties.name);
```

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -524,6 +538,427 @@ namespace geopm
+                metric_name == "XVE_STALL") {
+                for (unsigned int report_idx = 0; report_idx < num_reports; report_idx++) {
+                    // Each report contains num_metric entries of data.  We need to access
+                    // this metric (metric_idx) from each report ( report_idx * num_metric)
```
nit:
```suggestion
                    // this metric (metric_idx) from each report (report_idx * num_metric)
```

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -524,6 +538,427 @@ namespace geopm
+
+        for (unsigned int metric_idx = 0; metric_idx < num_metric; metric_idx++)
+        {
+            //TODO: It is possible that simply parsing all the metrics is
```
TODO for this PR?

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -524,6 +538,427 @@ namespace geopm
+                        ": LevelZero Event Create failed",
+                        __LINE__);
+
+        // TODO: nothing guarantees CHIP 0 was called first
```
:o

avilcheslopez commented on service/src/LevelZero.cpp:
```diff
@@ -524,6 +538,427 @@ namespace geopm
+        zet_metric_streamer_desc_t metric_streamer_desc = {
+            ZET_STRUCTURE_TYPE_METRIC_STREAMER_DESC,
+            nullptr,
+            4, // number of reports to notify on.  Targeting 4 reports, 1 per millisecond as that will, generally, 
```
This would be better to place in a symbolic constant, to make it easier to update when needed (as I know you've had to test different values), and make the code easier to read.

avilcheslopez reviewed (commented):
I'm still seeing multiple "TODO" comments, which makes it seem like this is still in draft form. Which of those TODOs need to be addressed as part of this PR?

avilcheslopez committed:
Fix zet branch build and unit test failures

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Fix typos in LevelZero.cpp

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Don't throw when reading level zero METRIC signals

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Fix minor issues after rebase

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Rebase and fix

- Fix build issues
- Fix LevelZeroIOGroup unit tests
- Fix memory leak: ZE context was not being destroyed

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Rebase and fix build issues

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

bgeltz commented:
Closing in favor of #4052
